### ISSUE TITLE
Understanding slow http_client output

### ISSUE THREAD

**Issue by ZachTB123** [benthosdev/benthos#2371]
Hello,

I'm in the process of doing some performance tests with Benthos. I'm following https://www.benthos.dev/docs/guides/performance_tuning as I go through the process.

This is my current config:

```yaml
http:
    address: 0.0.0.0:4195
    enabled: true
    root_path: /benthos

logger:
    level: INFO
    format: json
    add_timestamp: true
    static_fields:
        a_bunch: of_static_fields
        
tracer:
    open_telemetry_collector:
        http:
            - secure: false
              url: my-collector:4318
        tags:
            a_bunch: of_static_tags

metrics:
    mapping: |
        meta a_bunch = "of_static_tags"
    prometheus: {}

input:
    broker:
        copies: 1
        inputs:
            - kafka_franz:
                consumer_group: my-group
                seed_brokers:
                    - ${KAFKA_BOOTSTRAP_SERVERS}
                start_from_oldest: false
                tls:
                    client_certs:
                        - cert_file: /path/to/kafka-cert.pem
                          key_file: /path/to/kafka-key.pem
                    enabled: true
                    root_cas_file: /path/to/kafka-ca.crt
                topics:
                    - my-topic
                batching:
                  count: 100
                  period: 100ms
              label: subscriber_input

pipeline:
  processors:
    - json_schema:
        schema_path: file:////path/to/schema.json
      label: schema
    - catch:
        - label: row_filter_validation_failed
          mapping: root = deleted()
    - bloblang: |
         root = this
         root.data.metadata.ingestedTime = meta("kafka_timestamp_unix").number().ts_format(format: "2006-01-02T15:04:05Z", tz: "UTC")

output:
  fallback:
    - http_client:
        headers:
          Content-Type: application/json
          traceparent: ${! json("traceparent") }
        oauth2:
          client_key: client-id
          client_secret: ${WEBHOOK_OAUTH2_CLIENT_SECRET}
          enabled: true
          endpoint_params:
            grant_type:
              - client_credentials
          scopes:
            - my-scope
          token_url: https://my-token-url.com/oauth2/token
        url: https://my-url.com
        verb: POST        
      label: subscriber_output
  label: fallback_output           
```

Note that I'm using the code that's currently in main because I was encountering:
<img width="922" alt="Screenshot 2024-02-08 at 4 13 09 PM" src="https://github.com/benthosdev/benthos/assets/17228033/2df64b59-36a9-4ebe-b3c3-a53e86071de3">
which I believe was fixed by https://github.com/benthosdev/benthos/commit/abbcacb0a7560cd6ac334a725a18131e16da72d0. Oddly enough that broke my `json_schema` processor. I was previously using `schema_path: file:///path/to/schema.json` but had to change to `schema_path: file:////path/to/schema.json` because it was saying the file at `path/to/schema.json was not found`.

I'm trying to tweak my configuration because the end to end latency is too high. I'm basing this off of the traces I'm collecting.

This is the reported latency distribution:

<img width="1496" alt="Screenshot 2024-02-08 at 8 26 51 PM" src="https://github.com/benthosdev/benthos/assets/17228033/9f36521e-7168-4d13-bc6e-536758b04675">

I'm trying to track down where the latency is coming from because on the receiving webhook, the latency distribution for the requests it receives is under 1 second.

If I look at a single request on the Benthos side, the span for the output is high while the http request itself is low. For example:

<img width="1466" alt="Screenshot 2024-02-08 at 8 13 29 PM" src="https://github.com/benthosdev/benthos/assets/17228033/2ec88687-0036-4ffd-af97-471624c3df5d">

What could be the cause of the additional ~11 seconds of the `output_http_client` span if the request itself is only taking < 1 second?

Thanks

**Comment by Jeffail** [comment]
Hey @ZachTB123, thanks for mentioning the jsonschema issue, it should be fixed with https://github.com/benthosdev/benthos/commit/8a488565ca56ee45a7f4d4e28216052819367228.

Have you tried replacing the kafka_franz input with `generate` to get some test messages through? Once the request is successful the next thing is acknowledging the message and finalising the transaction, which includes registering the message checkpoint for commit. This shouldn't be blocked on anything that the kafka input is doing but it'd be good to rule it out.

**Comment by Jeffail** [comment]
Oh hang on I think I know what's happening, the latency is a bit skewed because it covers the entire batch of messages, so in your case you're probably looking at a trace of a message thats one of a batch of 100 or so and its output trace isn't finished until the entire batch is delivered.

You aren't actually benefitting from batching right now because https://www.benthos.dev/docs/components/outputs/http_client#batch_as_multipart isn't enabled and you aren't archiving the data to reduce the number of requests. I'd say either eliminate the batches by using a `split` processor to break those messages out, don't do any batching at all, and then maybe try increasing https://www.benthos.dev/docs/components/outputs/http_client#max_in_flight.

**Comment by twmb** [comment]
Jeffail diagnosed this as a batching configuration issue — the batch latency covers the entire batch, and without `batch_as_multipart` or archiving, batching wasn't providing benefit. Closing as answered. If you encounter further performance issues, feel free to open a new issue or PR.
